### DISCUSSION TITLE
How to have Milvus to load collection/partitions faster?

### DISCUSSION THREAD

**Discussion by xiaobingxia-at** [milvus-io/milvus#37393] [Q&A and General discussion] [answered]
Hello Milvus Experts, right now given 1.5-million-vector collection, it takes ~30 seconds to load the collection (client.load_collection(...)) into memory of query nodes. Just curious, is there anyway to accelerate this? like using larger query nodes? or any configurations we can tune?
I'm also wondering how long will it take if the collection has 30-million-vector in the future. Is it going to be many minutes?

I'm using IVF-PQ, (nlist = 128, m = 16)

**Comment by xiaofan-luan** [comment, marked as answer]
load it usually a one time task so once it's loaded it should be kept in memory or local ssd (if diskann used or mmap used).

you need to offer logs for us to analyze what's going on in your cluster and we can more advice, the bottleneck could be:

1. Minio bandwidtch -> if use S3 that won't be a problem
2. Querynode don't have enough concurrency or size(maybe local disk need to be faster) -> Here's our internal test, 1.5million load with 2core takes 15s, 18 million 24 core takes 60s(Still under optimization). And we are pretty sure there is something we can improve especially on large dataset. Usually the cluster it is, the faster the load could be. So if you create a larger cluster with tens of nodes, the loading process is usually less than 60s.

**Reply by xiaobingxia-at** [reply]
Thank you so much! I'm using S3 in this case. I'm using 4 query nodes, each query node is 30 Cores + 128GB memory. I don't use disk ANN, so I assume everything is loaded from s3 to memory?
In my benchmark, I do have 10 threads trying to do loading collection at the same time, and in the beginning it could be 180 seconds long to load a collection with 1.5-million 1536-dim vectors. But after a while, I guess not every thread is loading at the same time, the loading collection latency can reduce to 60-70 seconds long. 
So I think the many threads loading at the same time is one important reason why it is slow?

**Reply by xiaobingxia-at** [reply]
Also, can you elaborate more about "query node doesn't have enough concurrency"? Is there any configuration to tune query node to load data with more parallel threads?

**Reply by xiaofan-luan** [reply]
there is a pool on querynode use to load data.
Adding the pool thread number could potentially help the issue. the pool number is related to cpu cores of your docker

**Reply by xiaobingxia-at** [reply]
I'm looking at https://github.com/milvus-io/milvus/blob/master/configs/milvus.yaml
I only see one pool configuration for queryNode:
workerPooling:
    size: 10 # the size for worker querynode client pool
Is this the one?

**Comment by xiaobingxia-at** [comment]
I just realized that I was doing the "flush()" every time I do the upserts. Will that create a lot of small index files and impact the loading speed? Does the compaction process help remove the "small index file" problem and then speed up the loading ?

**Reply by xiaofan-luan** [reply]
usually you don't really need to do any manual flush. This is not recommended

**Reply by xiaofan-luan** [reply]
all the data can be seen even if don't do flush.
You only do flush when you are pretty sure this collection is not gonna to be written in a period
